### ISSUE TITLE
[JIT] lack of type support in tensor indexing.

### ISSUE THREAD

**Issue by chenbohua3** [pytorch/pytorch#38962]
Currently, variable `values` of `aten::index_put` only supports `torch.Tensor`. If we have such function:
``` python
@torch.jit.script
def f(x: torch.Tensor):
    mask = torch.ones(x.shape, dtype=torch.bool)
    x[mask] = 2
```
we will get:
```
RuntimeError:
Arguments for call are not valid.
The following variants are available:

  aten::index_put_(Tensor(a!) self, Tensor?[] indices, Tensor values, bool accumulate=False) -> (Tensor(a!)):
  Expected a value of type 'Tensor' for argument 'values' but instead found type 'int'.

  aten::index_put_(Tensor(a!) self, Tensor[] indices, Tensor values, bool accumulate=False) -> (Tensor(a!)):
  Expected a value of type 'List[Tensor]' for argument 'indices' but instead found type 'List[Optional[Tensor]]'.

The original call is:
  File "<ipython>", line 3
def f(x: torch.Tensor):
    mask = torch.ones(x.shape, dtype=torch.bool)
    x[mask] = 2
    ~~~~~~~~~~~ <--- HERE
```
As far as i know, there exist two workarounds:
``` python
x[mask] = torch.tensor(2)
# or
x.masked_fill_(mask, 2)
```
Both of the two workarounds are not as intuitive as `x[mask] = 2`. And if we are **under the context of multi-devices**, the former workaround becomes less concise because we may need to explicitly specify the tensor device.

So is there any plans to add more types support in `aten::index_put` to make the jit-capable indexing more intuitive:)
 

cc @suo

**Comment by eellison** [comment]
Hi @chenbohua3, thanks for the detailed issue and succinct repro! This was recently added in https://github.com/pytorch/pytorch/pull/38378 and works on master, so I'm closing.
